### PR TITLE
Annotation, simplifications, and renaming

### PR DESCRIPTION
**Pull request by MattWindsor91**
This large PRQ non-exhaustively does the following:

1) Annotates most of the top-level definitions in Starling with types (I didn't quite get everything covered);
2) Moves some definitions to nest inside top-level definitions;
3) Replaces `ViewDef` with `ViewDefiner` a modified `FuncTable` (now called `FuncDefiner`);
4) Kills off some type synonyms;
5) Generally refactors various parts of Starling.

### CODE REVIEW

File: Command.fs [hunk 1]
```diff
@@ -135,7 +135,7 @@ module Compose =
     ///     The next available intermediate stage number.
     ///     If the expression has no intermediate stages, we return 0.
     /// </returns>
-    and nextBoolIntermediate =
+    and nextBoolIntermediate : BoolExpr<Sym<_>> -> bigint =
```
Comment (bensimner) [inline Command.fs:138]: How exactly does F# interpret the _ in a type?

Is it equivalent to a (forall a. a) type or something?

Comment (MattWindsor91) [reply Command.fs:138]: I think it just asks the type inferrer to kick in at that point.  Really `_` is evidence either of laziness on my part, or me thinking that it literally does not matter what the type is there.

There are some annotations where I've used `_` to mean 'this doesn't change', which isn't quite right.

**Comment by septract** [comment]
Looks good but there are conflicts with master. @CaptainHayashi could you fix these?

**Comment by MattWindsor91** [comment]
@septract On it, might take a while (these are fairly nontrivial!).

**Comment by MattWindsor91** [comment]
@septract I've rebased on top of master (you might have to `pull --force` to check out locally).  This should mean there aren't any conflicts.

**Pull request merged**